### ISSUE TITLE
Large stack frames cause the listing to lag

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#2351]
**Describe the bug**
Large function stack frames, about 0x13f0 in size and roughly 1276 types, cause the listing to lag while scrolling.

**Expected behavior**
Gotta go fast

**Environment (please complete the following information):**
 - OS: Microsoft Windows [Version 10.0.19041.508]
 - Java Version: 15
 - Ghidra Version: 9.2_DEV f7f366b261589a65696c17852812566b9e6659bb (sorta)

**Additional context**
There really isn't much I can provide for this. I've only encountered it in large C++ programs that are subject to copyright.

Edit: Exactly how large of a function you ask?
`Exception while decompiling xxxxxxxx: Decompiler results exceeded payload limit of 50 MBytes`

**Comment by dragonmacher** [comment]
We have seen the behavior before when a function has an excessive number of xrefs.   In your case, with the large stack frame, I'm assuming your have many xrefs for the stack variables.  Is that the case?

If you have the time, energy and desire, it would be helpful to get some insight into what the slowness is.  I examine issues like this using `visualvm` that comes with the jdk.    I would start Ghidra and visualvm.  Once you connect to Ghidra in visualvm, then trigger the slow scrolling and see where visualvm indicates the time was consumed.

**Comment by astrelsky** [comment]
> We have seen the behavior before when a function has an excessive number of xrefs. In your case, with the large stack frame, I'm assuming your have many xrefs for the stack variables. Is that the case?
> 
> If you have the time, energy and desire, it would be helpful to get some insight into what the slowness is. I examine issues like this using `visualvm` that comes with the jdk. I would start Ghidra and visualvm. Once you connect to Ghidra in visualvm, then trigger the slow scrolling and see where visualvm indicates the time was consumed.

Unfortunately visualvm doesn't support jdk15 yet but I found something else that I could get a free trial for.

Apparently this is where the problem starts.
https://github.com/NationalSecurityAgency/ghidra/blob/3c683ae770cf495deb32588202b00b88e125a4a4/Ghidra/Features/Base/src/main/java/ghidra/app/util/viewer/format/FieldFormatModel.java#L502

Most of the slow down is being caused by `db.KeyToRecordIterator.next()` with most of the time being spent here
https://github.com/NationalSecurityAgency/ghidra/blob/3c683ae770cf495deb32588202b00b88e125a4a4/Ghidra/Framework/DB/src/main/java/db/Schema.java#L268

You could create a corresponding `Supplier<Field>[]` which would remove any overhead of repeatedly constructing it via reflection. Alternatively you could also get away with a static map which would map each Field Class to its corresponding new.

[Call-tree-–-By-thread.html.txt](https://github.com/NationalSecurityAgency/ghidra/files/5379718/Call-tree-.-By-thread.html.txt)

**Comment by dragonmacher** [comment]
Thanks for the dump.  It shows what I suspected, that the xrefs are doing too much work for large functions (look at the section around the one you mentioned, where it points to the `VariableXRefHeaderFieldFactory`.  This is loading all symbols to figure out how to display xrefs in the function header.  The problem with the code is that it does not place any limitation on the number of symbols retrieved.

It needs to be smarter.  For example, it could understand that the UI is set to display a limited number of xrefs, depending upon the option, and to then stop loading symbols at that point.  For example, if the option indicates to show 5 xrefs, then stop loading xrefs after 5 are reached.   This may not be enough if the volume of the stack members is huge.  Some sort of caching may also be required. 

We haven't fixed this yet, as it only affects a small number of binaries, such as what you are using.

**Comment by astrelsky** [comment]
It would be nice if the amount of the numbers in the stack frame which gets displayed is limited as well. It is about as fun as you could imagine to scroll through.

I did do the following test just to have a look and while it did improve I'm not entirely certain if it is worth making the change to it. This was done about as lazily as possible.

```java
private static Map<Class<?>, Supplier<Field>> fieldMap;

static {
	fieldMap = new HashMap<>();
	fieldMap.put(BinaryField.class, BinaryField::new);
	fieldMap.put(BooleanField.class, BooleanField::new);
	fieldMap.put(ByteField.class, ByteField::new);
	fieldMap.put(IntField.class, IntField::new);
	fieldMap.put(LongField.class, LongField::new);
	fieldMap.put(ShortField.class, ShortField::new);
	fieldMap.put(StringField.class, StringField::new);
}

private static Field newField(Class<?> clazz) throws Exception {
	Supplier<Field> supplier = fieldMap.get(clazz);
	return supplier != null ? supplier.get() :
		(Field) clazz.getDeclaredConstructor().newInstance();
}
```

I then replaced all uses of Class.newInstance with the new method.

While it's not a crazy amount it is still a speed boost of around 80% on a bit of code used everywhere in Ghidra.
[Live-call-tree.html.txt](https://github.com/NationalSecurityAgency/ghidra/files/5380567/Live-call-tree.html.txt)

**Comment by dragonmacher** [comment]
I created a fork and invited you to examine changes related to this ticket in a branch named 'GH-2351-dragonmacher-slow-listing-fix'

**Comment by astrelsky** [comment]
> I created a fork and invited you to examine changes related to this ticket in a branch named '[GH-2351](https://github.com/NationalSecurityAgency/ghidra/issues/2351)-dragonmacher-slow-listing-fix'
> 
> Please try the changes in your problem binary.

Will do.

**Comment by astrelsky** [comment]
It is good. I ran the profiler in case there was a measurable problem somewhere but I don't see anything that stands out. I also held the pg dn key for a good 5 minutes and passed a handful of stack frames which were several times larger than the one I was initially using to test the issue and had no problem. I've attached that data just because.

While not related I came across a list of thrown exceptions and happened to notice that `docking.DockingActionProxy.createMenuItem(boolean) DockingActionProxy.java:178` had thrown 3298 NoSuchFieldException 2442 of which were from `docking.menu.MenuBarManager.getMenuBar() MenuBarManager.java:142`. It's not relevant but am bringing it up incase it should be looked into.

[Live-call-tree.html.txt](https://github.com/NationalSecurityAgency/ghidra/files/5393894/Live-call-tree.html.txt)
[Live-call-tree.csv.txt](https://github.com/NationalSecurityAgency/ghidra/files/5393895/Live-call-tree.csv.txt)

Held
-----

[Live-call-tree-held.zip](https://github.com/NationalSecurityAgency/ghidra/files/5393930/Live-call-tree-held.zip)
[Live-call-tree-held.csv.txt](https://github.com/NationalSecurityAgency/ghidra/files/5393931/Live-call-tree-held.csv.txt)
